wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala

```diff
@@ -268,7 +268,22 @@ private[wallet] trait TransactionProcessing extends WalletLogger {
               s"Failed to update confirmation count of transaction=${transaction.txIdBE}",
               err))
 
-          updateF
+          updateF.flatMap { updated =>
```

Comment:
So this is where we can get a lot of leverage out of using ADTs. With how the current code is written you are using `if()` which doesn't give you any compiler guarantees about handling states correctly, _and_ doesn't give us any warnings if someday we need to add more types to `TxoState`. Here is a more functional and scalable approach to this:

```scala
          for {
            updated <- updateF
            confsOpt <- chainQueryApi.getNumberOfConfirmations(blockHash)
            stateChange <- {
              confsOpt match {
                case None => Future.successful(updated)
                case Some(confs) =>
                  if (confs >= walletConfig.requiredConfirmations) {
                    updated.state match {
                      case pending @ (TxoState.PendingConfirmationsReceived | TxoState.PendingConfirmationsSpent) =>
                        spendingInfoDAO.update(
                          updated.copyWithState(pending))
                      case TxoState.DoesNotExist =>
                        // ??? corrupt chain ???
                        ???
                      case TxoState.ConfirmedReceived | TxoState.ConfirmedSpent =>
                        // ??? reorg ???
                        ???
                    }
                  } else {
                    //verify that the state is at least Pending ??
                    ???
                  }
              }
            }
          } yield stateChange
```

With ADTs we can more accurately model what states are actually possible, if we decide to use them, and it gives us a clear framework for us to think in for what we possibly need to handle. When in doubt about ADTs, [listen to Jane Street](https://youtu.be/FnBPECrSC7o?t=1879)! 

Honestly, even my code is a little ugly. We should probably break out the pattern match into it's own helper method.

Reply:
Also it looks like on my first iteration of writing this code i forgot a state... just to show the power of ADTs I didn't consider `Reserved`! Doh. 

>[warn] /home/suredbits/dev/bitcoin-s/wallet/src/main/scala/org/bitcoins/wallet/internal/TransactionProcessing.scala:279:29: match may not be exhaustive.
[warn] It would fail on the following input: Reserved
[warn]                     updated.state match {